### ISSUE TITLE
NIO can’t assume it can handle all 108 bytes of sockaddr_un

### ISSUE THREAD

**Issue by Lukasa** [apple/swift-nio#1770]
Right now we assume that we can just move a sockaddr_un around as a typed struct. That’s not safe: the type may be arbitrarily short as well as arbitrarily long. 

We should investigate fixing up SocketAddress to tolerate this. 

See also: nix-rust/nix#177

**Comment by weissi** [comment]
I don't _think_ this is correct for the two platforms we officially support. Both Linux and macOS have fixed-sized char arrays at the back of a `sockaddr_un`. We do (intentionally) not support paths > 104 (IIRC) bytes because that's the maximum that will fix in either system.

Darwin:

```
  1> import Darwin
  2> MemoryLayout<sockaddr_un>.size
$R0: Int = 106
```

```
struct  sockaddr_un {
	unsigned char   sun_len;        /* sockaddr len including null */
	sa_family_t     sun_family;     /* [XSI] AF_UNIX */
	char            sun_path[104];  /* [XSI] path name (gag) */
};
```

(glibc) Linux:

```
  2> MemoryLayout<sockaddr_un>.size
$R0: Int = 110
```

```
# grep -A5 sockaddr_un /usr/include/x86_64-linux-gnu/sys/un.h
struct sockaddr_un
  {
    __SOCKADDR_COMMON (sun_);
    char sun_path[108];		/* Path name.  */
  };

```

am I misunderstanding what you're saying @Lukasa ?

**Comment by weissi** [comment]
the hack that we don't support is this one:

```
struct sockaddr_un *long_path = malloc(sizeof(*long_path) + 1024);
*long_path = {0};
memcpy(long_path.sun_path, "/very/long/path/........")
```

if you hand that to NIO, it should reject it as path too long.

**Comment by Lukasa** [comment]
Yeah, you're misunderstanding what I'm saying. Consider this code:

```swift
import Glibc
import Glibc

var sockets: [CInt] = [0, 0]
precondition(
    socketpair(AF_UNIX, CInt(SOCK_STREAM.rawValue), 0, &sockets) == 0,
    "Failed to create socketpair"
)

var ss = sockaddr_storage()
let length: socklen_t = withUnsafeMutablePointer(to: &ss) { storagePtr in
    return storagePtr.withMemoryRebound(to: sockaddr.self, capacity: 1) { reboundPtr in
        var length = socklen_t(MemoryLayout<sockaddr_storage>.size)
        let rc = getpeername(sockets[0], reboundPtr, &length)
        precondition(rc == 0, "failed call to getpeername")
        return length
    }
}
precondition(ss.ss_family == AF_UNIX)
precondition(
    length == MemoryLayout<sockaddr_un>.size,
    "Unexpected length for sockaddr_un: \(length)"
)
```

This code is a Swift translation of the nix-rust issue linked above. In it we create a socketpair and then call getpeername on it and check the result.

The above code crashes in Swift with `Precondition failed: Unexpected length for sockaddr_un: 2: file test/test.swift, line 19`. Now, for the above code this isn't an issue: we allocated the buffer and we know it's large enough. But we can tweak the above program to:

```swift
import Glibc

var sockets: [CInt] = [0, 0]
precondition(
    socketpair(AF_UNIX, CInt(SOCK_STREAM.rawValue), 0, &sockets) == 0,
    "Failed to create socketpair"
)

var ss: (sa_family_t, UInt8, UInt8, UInt8) = (0, 0, 0, 0)
let length: socklen_t = withUnsafeMutablePointer(to: &ss) { storagePtr in
    return storagePtr.withMemoryRebound(to: sockaddr.self, capacity: 1) { reboundPtr in
        var length = socklen_t(MemoryLayout<sa_family_t>.size + 3)
        let rc = getpeername(sockets[0], reboundPtr, &length)
        precondition(rc == 0, "failed call to getpeername")
        return length
    }
}
precondition(ss.0 == AF_UNIX)
precondition(
    length == MemoryLayout<sockaddr_un>.size,
    "Unexpected length for sockaddr_un: \(length)"
)
```

Here we no longer use `sockaddr_storage` but an insane tiny buffer. This still works, and crashes at the same precondition location where `length == 2`.

Importantly, this implies that it is totally possible to have a Linux `sockaddr_un` whose valid bytes are _shorter_ than the size of that structure. The Linux kernel and glibc don't care: they assume that `sockaddr_un` is basically a flexible array member. Indeed, the manual page (`man 7 unix`) says this:

```
      Applications that retrieve socket addresses can (portably) code
       to handle the possibility that there is no null terminator in
       sun_path by respecting the fact that the number of valid bytes in
       the pathname is:

           strnlen(addr.sun_path, addrlen - offsetof(sockaddr_un,
       sun_path))
```

Note that while this is talking about null-terminators, it says _the number of valid bytes in the pathname_. In the case of this code, the number of valid bytes in the pathname is zero.

**Comment by Lukasa** [comment]
To be clear, the reason it matters is that we have this initializer here:

https://github.com/apple/swift-nio/blob/d22d89804cd42cc7c1f938d588c02ffc1cec5497/Sources/NIO/SocketAddresses.swift#L265-L267

This allows any arbitrary `sockaddr_un` to be passed in, and so needs to be coded defensively.

**Comment by weissi** [comment]
@Lukasa oh yes, we need to check that the string is short enough there (or just `0` the last byte and piggy-back on the follow-up error because we can't fail in that `init`)

**Comment by Lukasa** [comment]
We also have to not copy more bytes than the string actually holds, or we risk faulting.

**Comment by weissi** [comment]
@Lukasa yes, that's why I meant just `0`ing the last byte (in the structure).

**Comment by Lukasa** [comment]
That's not what I mean, I mean the structure being passed to us may literally not be the right length. Thinking on this more, this API is very difficult to hold safely from Swift. I think the best move is going to be to try to remove all of the sockaddr_un APIs entirely tbqh.

**Comment by weissi** [comment]
> That's not what I mean, I mean the structure being passed to us may literally not be the right length. Thinking on this more, this API is very difficult to hold safely from Swift. I think the best move is going to be to try to remove all of the sockaddr_un APIs entirely tbqh.

IIRC, back in the day we decided to support a certain maximum path length. That was chosen carefully such that it just fits into the smaller of the `sockaddr_un`s on the OSes we support. So if the structure "is just the right length", IMHO it's one too long because the trailing `0` byte is part of the path... So to fix it, we can just write `0` to the last byte (if we don't do that already). But yeah, also happy deprecating that init, should still make sure it's memory safe.

**Comment by weissi** [comment]
> That's not what I mean, I mean the structure being passed to us may literally not be the right length.

How do you mean? I can't be shorter than `sizeof(struct sockaddr_un)`, that'd be illegal.

**Comment by weissi** [comment]
And if we wanted to support that, then we also could because if it is shorter, then there must be a zero byte that we can scan for safely.

**Comment by Lukasa** [comment]
> How do you mean? I can't be shorter than sizeof(struct sockaddr_un), that'd be illegal.

Be that as it may, Linux is quite happy to work with such a structure.

**Comment by weissi** [comment]
Ok, we discussed this offline:

- Right now, we're correct but user hostile. We accept a `sockaddr_un` (note: _not_ a `sockaddr_un *`) so if a user got an (illegal) short `sockaddr_un *`, they would probably do `SocketAddress(pointerToSockAddrUn.pointee)` which is illegal (but looks totally fine).
- In the future, we shouldn't be as user hostile and accept a `sockaddr_un *` and transform it into our own representation (which would hold the path in `[UInt8]` or `String` and can therefore tolerate arbitrarily sized paths). When we do have to call the kernel, we'd transparently transform to an appropriately sized `sockaddr_un *`.
